rust/src/mqtt/detect.rs

```diff
@@ -17,421 +17,1532 @@
                 *buffer_len = p.len() as u32;
-                return 1;
+                return true;
             }
         }
     }
 
     *buffer = ptr::null();
     *buffer_len = 0;
-    return 0;
-}
-
-#[no_mangle]
-pub unsafe extern "C" fn rs_mqtt_tx_get_connack_sessionpresent(
-    tx: &MQTTTransaction, session_present: *mut bool,
-) -> u8 {
-    for msg in tx.msg.iter() {
-        if let MQTTOperation::CONNACK(ref ca) = msg.op {
-            *session_present = ca.session_present;
-            return 1;
-        }
-    }
-    return 0;
+    return false;
 }
 
-#[no_mangle]
-pub unsafe extern "C" fn rs_mqtt_tx_get_publish_topic(
-    tx: &MQTTTransaction, buffer: *mut *const u8, buffer_len: *mut u32,
-) -> u8 {
+unsafe extern "C" fn mqtt_tx_get_publish_topic(
+    tx: *const c_void, _flags: u8, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> bool {
+    let tx = cast_pointer!(tx, MQTTTransaction);
     for msg in tx.msg.iter() {
         if let MQTTOperation::PUBLISH(ref pubv) = msg.op {
             let p = &pubv.topic;
             if !p.is_empty() {
                 *buffer = p.as_ptr();
                 *buffer_len = p.len() as u32;
-                return 1;
+                return true;
             }
         }
     }
 
     *buffer = ptr::null();
     *buffer_len = 0;
-
-    return 0;
+    return false;
 }
 
-#[no_mangle]
-pub unsafe extern "C" fn rs_mqtt_tx_get_publish_message(
-    tx: &MQTTTransaction, buffer: *mut *const u8, buffer_len: *mut u32,
-) -> u8 {
+unsafe extern "C" fn mqtt_tx_get_publish_message(
+    tx: *const c_void, _flags: u8, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> bool {
+    let tx = cast_pointer!(tx, MQTTTransaction);
     for msg in tx.msg.iter() {
         if let MQTTOperation::PUBLISH(ref pubv) = msg.op {
             let p = &pubv.message;
             if !p.is_empty() {
                 *buffer = p.as_ptr();
                 *buffer_len = p.len() as u32;
-                return 1;
+                return true;
             }
         }
     }
 
     *buffer = ptr::null();
     *buffer_len = 0;
+    return false;
+}
 
+fn mqtt_tx_get_reason_code(tx: &MQTTTransaction) -> Option<u8> {
+    for msg in tx.msg.iter() {
+        match msg.op {
+            MQTTOperation::PUBACK(ref v)
+            | MQTTOperation::PUBREL(ref v)
+            | MQTTOperation::PUBREC(ref v)
+            | MQTTOperation::PUBCOMP(ref v) => {
+                if let Some(rcode) = v.reason_code {
+                    return Some(rcode);
+                }
+            }
+            MQTTOperation::AUTH(ref v) => {
+                return Some(v.reason_code);
+            }
+            MQTTOperation::CONNACK(ref v) => {
+                return Some(v.return_code);
+            }
+            MQTTOperation::DISCONNECT(ref v) => {
+                if let Some(rcode) = v.reason_code {
+                    return Some(rcode);
+                }
+            }
+            _ => return None,
```

Comment:
@satta why do we not continue with the loop ?

Reply:
Right, we should. Good catch! This is likely an oversight when extending transactions to multi-operations.

Reply:
will fix in next version

Reply:
Thanks!